src/gt4py/analysis/passes.py

```diff
@@ -1290,6 +1290,89 @@ def apply(cls, transform_data: TransformData) -> None:
         cls.DemoteSymbols.apply(transform_data.implementation_ir, demotables)
 
 
+class ReduceTemporaryStoragesPass(TransformPass):
+    """Demote 3D temporaries only used within a single stage to 2D temporaries."""
+
+    class ReducibleFieldsCollector(gt_ir.IRNodeVisitor):
+        @classmethod
+        def apply(cls, node: gt_ir.StencilImplementation) -> Set[str]:
+            collector = cls()
+            return collector(node)
+
+        def __call__(self, node: gt_ir.StencilImplementation) -> Set[str]:
+            assert isinstance(node, gt_ir.StencilImplementation)
+            self.interval: gt_ir.AxisInterval = None
+            self.iteration_order: gt_ir.IterationOrder = None
+            self.reduced_fields: Dict[str, gt_ir.AxisInterval] = {
+                temp_field: None for temp_field in node.temporary_fields
+            }
+            self.visit(node)
+            return set(self.reduced_fields.keys())
+
+        def visit_MultiStage(self, node: gt_ir.MultiStage) -> None:
+            self.iteration_order = node.iteration_order
+            self.generic_visit(node)
+
+        def visit_ApplyBlock(self, node: gt_ir.ApplyBlock) -> None:
+            self.interval = node.interval
+            self.generic_visit(node)
+
+        def visit_FieldRef(self, node: gt_ir.FieldRef) -> None:
+            field_name = node.name
+            if field_name in self.reduced_fields:
+                if self.iteration_order == gt_ir.IterationOrder.PARALLEL:
+                    # Do not reduce fields accessed from parallel k-intervals
+                    self.reduced_fields.pop(field_name)
+                else:
+                    offsets: List[int] = list(node.offset.values())
+                    if offsets[-1] != 0:
```

Comment:
could use a check if "K" is in storage at all here, for additional safety e.g. if calling repeatedly

Reply:
Implemented!